lib/Dialect/Arc/Transforms/LowerStateRewrite.cpp

```diff
@@ -0,0 +1,1206 @@
+  edge = builder.create<comb::AndOp>(loc, edge, clock);
+  return edge;
+}
+
+/// Get the builder appropriate for the given phase.
+OpBuilder &ModuleLowering::getBuilder(Phase phase) {
+  switch (phase) {
+  case Phase::Initial:
+    return initialBuilder;
+  case Phase::Old:
+  case Phase::New:
+    return builder;
+  case Phase::Final:
+    return finalBuilder;
+  }
+}
+
+/// Get the lowered value, or emit a diagnostic and return null.
+Value ModuleLowering::requireLoweredValue(Value value, Phase phase,
+                                          Location useLoc) {
+  if (auto lowered = loweredValues.lookup({value, phase}))
+    return lowered;
+  auto d = emitError(value.getLoc()) << "value has not been lowered";
+  d.attachNote(useLoc) << "value used here";
+  return {};
+}
+
+//===----------------------------------------------------------------------===//
+// Operation Lowering
+//===----------------------------------------------------------------------===//
+
+/// Create a new `scf.if` operation with the given builder, or reuse a previous
+/// `scf.if` if the builder's insertion point is located right after it.
+static scf::IfOp createOrReuseIf(OpBuilder &builder, Value condition,
+                                 bool withElse) {
+  scf::IfOp ifClockOp;
+  if (auto ip = builder.getInsertionPoint(); ip != builder.getBlock()->begin())
+    if (auto ifOp = dyn_cast<scf::IfOp>(*std::prev(ip)))
+      if (ifOp.getCondition() == condition)
+        return ifOp;
+  return builder.create<scf::IfOp>(condition.getLoc(), condition, withElse);
+}
+
+/// This function is called from the lowering worklist in order to perform a
+/// depth-first traversal of the surrounding module. These functions call
+/// `lowerValue` to mark their operands as dependencies in the depth-first
+/// traversal, and to map them to the lowered value in one go.
+LogicalResult OpLowering::lower() {
+  return TypeSwitch<Operation *, LogicalResult>(op)
+      // Operations with special lowering.
+      .Case<StateOp, sim::DPICallOp, MemoryOp, TapOp, InstanceOp, hw::OutputOp,
+            seq::InitialOp, llhd::FinalOp>([&](auto op) { return lower(op); })
+
+      // Operations that should be skipped entirely and never land on the
+      // worklist to be lowered.
+      .Case<MemoryWritePortOp, MemoryReadPortOp>([&](auto op) {
+        op.emitOpError() << "is handled by memory op and must be skipped";
+        return success();
+      })
+
+      // All other ops are simply cloned into the lowered model.
+      .Default([&](auto) { return lowerDefault(); });
+}
+
+/// Called for all operations for which there is no special lowering. Simply
+/// clones the operation.
+LogicalResult OpLowering::lowerDefault() {
+  // Make sure that all operand values are lowered first.
+  IRMapping mapping;
+  auto anyFailed = false;
+  op->walk([&](Operation *nestedOp) {
+    for (auto operand : nestedOp->getOperands()) {
+      if (op->isAncestor(operand.getParentBlock()->getParentOp()))
```

Comment:
Yes. We'll need to figure out how exactly nested stateful ops can be handled, for example when there is an `arc.clock_domain` op present, or something like an `arc.vectorize` or `arc.task` or `arc.group`. I'm not entirely sure how that will work, since the container op's semantics likely affects how the contained ops are lowered.